src/propcache/_helpers_c.pyx

```diff
@@ -4,6 +4,60 @@ from types import GenericAlias
 
 cdef _sentinel = object()
 
+
+cdef class CacheBase:
+    """Base class for objects that use cached properties.  This class
+    provides a _cache attribute that is used to store the results of
+    cached properties.
+
+    Callers are responsible for calling super().__init__() in their
+    __init__() methods to ensure that the _cache attribute is
+    initialized.
+    """
+
+    cdef public dict _cache
+
+    def __init__(self):
+        self._cache = {}
+
+
+cdef class base_cached_property:
+    """Use as a class method decorator.  It operates almost exactly like
+    the Python `@property` decorator, but it puts the result of the
+    method it decorates into the instance dict after the first call,
+    effectively replacing the function it decorates with an instance
+    variable.  It is, in Python parlance, a data descriptor. This version
+    requires that the base class CacheBase is used, which provides
+    the _cache attribute.
+
+    """
+
+    cdef readonly object wrapped
+    cdef object name
+
+    def __init__(self, wrapped):
+        self.wrapped = wrapped
+        self.name = wrapped.__name__
+
+    @property
+    def __doc__(self):
+        return self.wrapped.__doc__
+
+    def __get__(self, inst, owner):
+        if inst is None:
+            return self
+        val = (<CacheBase>inst)._cache.get(self.name, _sentinel)
```

Comment:
Problem is this is unsafe to cast.  If they don't pass the correct object it could result in a seg fault.  Will let the benchmark rerun and than check the cost of a type check

Reply:
changed it to check the type so should be ok now